subprojects/core-api/src/main/java/org/gradle/api/internal/file/FileSystemLocationPropertyInternal.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.file;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.file.FileSystemLocation;
+import org.gradle.api.file.FileSystemLocationProperty;
+import org.gradle.api.provider.Provider;
+
+/**
+ * This class was added just to support FileSystemLocation properties in the convention mapping.
+ *
+ * Should be removed once ConventionMapping is removed.
+ */
+@NonNullApi
+public interface FileSystemLocationPropertyInternal<T extends FileSystemLocation> extends FileSystemLocationProperty<T> {
```

Comment:
It seems the convention is to also annotate the public type with [`@HasInternalProtocol`](https://github.com/gradle/gradle/blob/master/platforms/core-runtime/java-language-extensions/src/main/java/org/gradle/internal/HasInternalProtocol.java#L33):

```
/**
 * Indicates that there is an internal complementary protocol to the public type that is annotated with this.
 *
 * This should only be used on a type that is always assumed to also implement the internal protocol by Gradle internals.
 *
 * This exists to help anyone reading the source code realise that there is an internal component to the type.
 */
 ```

Reply:
This annotation is for public types to warn users to not implement them. This type is internal anyway, so there is added value to add this annotation

Reply:
Sorry, I meant to annotate the public type `FileSystemLocationProperty`, not this internal protocol itself.

 > This annotation is for public types to warn users to not implement them.
 
Are you sure, Anže? The javadoc seems to describe it is meant exactly to denote the case we have here, a public type (`FileSystemLocationProperty`) has an internal protocol (`FileSystemLocationPropertyInternal`) that must be implemented by all implementations of the public type (`AbstractFileVar` being the common ancestor to all of them):

> This should only be used on a type that is always assumed to also implement the internal protocol by Gradle internals.

Reply:
That makes sense yes, I though you meant adding it to the new type. My bad

Reply:
Fixed :)